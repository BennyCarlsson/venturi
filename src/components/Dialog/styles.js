import styled from 'styled-components'
import Colors from 'tokens/colors'
import { motion } from 'framer-motion'

export const DialogWrapper = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  padding: 8px 8px 16px 8px;
  background-color: ${Colors.vBlue};
`

export const InnerWrapper = styled(motion.div)`
  position: relative;
  height: 100%;
  width: 100%;
  border-radius: 8px;
  background-color: ${Colors.white};
`
