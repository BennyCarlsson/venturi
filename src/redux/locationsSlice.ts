import { createSlice, createAsyncThunk } from '@reduxjs/toolkit'
import { getLocationOName } from 'api/vasttrafik/vasttrafik'
import { LocationList } from 'types'

const _getLocationOName = async (
  name: string,
  { signal }: { signal: AbortSignal }
) => {
  signal.addEventListener('abort', () => {
    console.log('cancelled')
  })
  return getLocationOName(name, signal)
}

export const fetchDestinationLocationsOnName = createAsyncThunk(
  'locations/fetchDestinationLocationsOnName',
  _getLocationOName
)

export const fetchOriginLocationsOnName = createAsyncThunk(
  'locations/fetchOriginLocationsOnName',
  _getLocationOName
)

export enum LocationType {
  StopLocation,
  CoordLocation,
}

export type Location = {
  id: string
  idx: string
  name: string
  lat: string
  lon: string
  type: LocationType
}

// Todo extract and write tests
const convertData = (data: LocationList): Location[] | undefined => {
  console.log('data', data)
  let stopLocations: Location[] = []
  if (data.StopLocation) {
    if (!Array.isArray(data.StopLocation)) {
      data.StopLocation = [data.StopLocation]
    }
    stopLocations = data.StopLocation?.map((loc) => ({
      id: loc.id,
      idx: loc.idx,
      name: loc.name,
      lat: loc.lat,
      lon: loc.lon,
      type: LocationType.StopLocation,
    }))
  }

  let coordLocation: Location[] = []
  if (data.CoordLocation) {
    if (!Array.isArray(data.CoordLocation)) {
      data.CoordLocation = [data.CoordLocation]
    }
    coordLocation = data.CoordLocation?.map((loc) => ({
      id: loc.name,
      idx: loc.idx,
      name: loc.name,
      lat: loc.lat,
      lon: loc.lon,
      type: LocationType.CoordLocation,
    }))
  }
  return stopLocations
    .concat(coordLocation)
    .sort((a, b) => parseInt(a.idx) - parseInt(b.idx))
}

export interface TripState {
  loadingOrigin: boolean
  originLocations?: Location[]
  originError?: string
  loadingDestination: boolean
  destinationLocations?: Location[]
  destinationError?: string
}

const initialState: TripState = {
  loadingOrigin: false,
  loadingDestination: false,
}

const locationSearchSlice = createSlice({
  name: 'location',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(fetchOriginLocationsOnName.pending, (state, action) => {
      state.loadingOrigin = true
    })
    builder.addCase(fetchOriginLocationsOnName.fulfilled, (state, action) => {
      state.loadingOrigin = false
      if (action.payload.LocationList.error) {
        state.originError = action.payload.LocationList.errorText
      } else {
        state.originLocations = convertData(action.payload.LocationList)
      }
    })
    builder.addCase(
      fetchDestinationLocationsOnName.pending,
      (state, action) => {
        state.loadingDestination = true
      }
    )
    builder.addCase(
      fetchDestinationLocationsOnName.fulfilled,
      (state, action) => {
        state.loadingDestination = false
        if (action.payload.LocationList.error) {
          state.destinationError = action.payload.LocationList.errorText
        } else {
          state.destinationLocations = convertData(action.payload.LocationList)
        }
      }
    )
  },
})

export default locationSearchSlice.reducer
