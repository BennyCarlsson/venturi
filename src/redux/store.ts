import { configureStore } from '@reduxjs/toolkit'
import locationsReducer from './locationsSlice'
import tripReducer from './tripSlice'
import tripListReducer from './tripListSlice'
import dialogReducer from './dialogSlice'
import { loadOriginAndDestination, loadTriplList } from '../utils/localStorage'

const localStorageDataTrip = loadOriginAndDestination()
const localStorageDataTripList = loadTriplList()

const store = configureStore({
  reducer: {
    trip: tripReducer,
    tripList: tripListReducer,
    locations: locationsReducer,
    dialog: dialogReducer,
  },
  preloadedState:
    localStorageDataTrip && localStorageDataTripList
      ? {
          trip: {
            origin: localStorageDataTrip.origin,
            destination: localStorageDataTrip.destination,
          },
          tripList: localStorageDataTripList,
        }
      : {},
})

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
export default store
