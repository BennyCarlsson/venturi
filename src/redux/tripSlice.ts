import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import { getTrip } from 'api/vasttrafik/vasttrafik';

type tripOrderType = {
  fromId: string;
  toId: string;
};
export const fetchTrip = createAsyncThunk('trip/fetchTrip', async (tripOrder: tripOrderType) => {
  const response = await getTrip(tripOrder.fromId, tripOrder.toId);
  return response;
});

export const tripSlice = createSlice({
  name: 'test',
  initialState: {
    value: 0
  },
  reducers: {
    increment: (state) => {
      // Redux Toolkit allows us to write "mutating" logic in reducers. It
      // doesn't actually mutate the state because it uses the Immer library,
      // which detects changes to a "draft state" and produces a brand new
      // immutable state based off those changes
      console.log('hello');

      state.value += 1;
    },
    decrement: (state) => {
      state.value -= 1;
    },
    incrementByAmount: (state, action) => {
      state.value += action.payload;
    }
  }
});

// Action creators are generated for each case reducer function
export const { increment, decrement, incrementByAmount } = tripSlice.actions;

export default tripSlice.reducer;
