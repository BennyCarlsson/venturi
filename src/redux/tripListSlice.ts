import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { Location } from './locationsSlice'

export interface TripListState {
  trips: { name: string; origin: Location; destination: Location }[]
}

const initialState: TripListState = { trips: [] }

const tripListSlice = createSlice({
  name: 'trip',
  initialState,
  reducers: {
    addTripToList: (
      state,
      action: PayloadAction<{
        name?: string
        origin: Location
        destination: Location
      }>
    ) => {
      state.trips.unshift({
        name: action.payload.name || action.payload.origin.name,
        origin: action.payload.origin,
        destination: action.payload.destination,
      })
    },
    moveTripToFirst: (state, action: PayloadAction<{ index: number }>) => {
      const index = action.payload.index
      const trip = state.trips[index]
      state.trips.splice(index, 1)
      state.trips.unshift(trip)
      return state
    },
  },
})

export const { addTripToList, moveTripToFirst } = tripListSlice.actions

export default tripListSlice.reducer
