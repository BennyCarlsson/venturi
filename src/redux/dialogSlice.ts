import { createSlice } from '@reduxjs/toolkit'
import { RootState } from './store'

export type DialogType = {
  show: boolean
  isIntro: boolean
}
const initialState: DialogType = {
  show: false,
  isIntro: false,
}
const dialogSlice = createSlice({
  name: 'dialog',
  initialState,
  reducers: {
    showDialog: (state) => {
      state.show = true
      state.isIntro = false
    },
    showIntroDialog: (state) => {
      state.show = true
      state.isIntro = true
    },
    hideDialog: (state) => {
      state.show = false
      state.isIntro = false
    },
  },
})

export const { showDialog, showIntroDialog, hideDialog } = dialogSlice.actions

export const selectShowDialog = (state: RootState) => state.dialog.show
export const selectIsIntro = (state: RootState) => state.dialog.isIntro
export default dialogSlice.reducer
