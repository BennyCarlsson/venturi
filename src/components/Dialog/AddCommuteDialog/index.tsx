import { useState } from 'react'
import ChooseDestionationView from './ChooseDestionationView'
import NameCommuteView from './NameCommuteView'
import { AddCommuteDialogWrapper } from './styles'
import { Location } from 'redux/locationsSlice'
import { useAppDispatch } from 'hooks/redux'
import { setCurrentTable } from 'redux/tripSlice'
import { saveOriginAndDestination } from 'utils/localStorage'
import { hideDialog } from 'redux/dialogSlice'
import { addTripToList } from 'redux/tripListSlice'
import IntroView from './IntroView'

type AddCommuteDialogProps = {
  isIntroView: boolean
}

const AddCommuteDialog = ({ isIntroView, ...props }: AddCommuteDialogProps) => {
  const dispatch = useAppDispatch()
  const [displayView, setDisplayView] = useState<
    'IntroView' | 'ChooseDestinationView' | 'NameCommuteView'
  >(isIntroView ? 'IntroView' : 'ChooseDestinationView')
  const [originLocation, setOriginLocation] = useState<Location | null>(null)
  const [name, setName] = useState('')
  const [destinationLocation, setDestinationLocation] =
    useState<Location | null>(null)

  const goToIntroView = () => {
    setDisplayView('IntroView')
  }

  const goToNameCommuteView = () => {
    setDisplayView('NameCommuteView')
  }

  const goToChooseDestinationView = () => {
    setDisplayView('ChooseDestinationView')
  }

  const createTrip = () => {
    if (originLocation && destinationLocation) {
      dispatch(
        setCurrentTable({
          name,
          origin: originLocation,
          destination: destinationLocation,
        })
      )
      saveOriginAndDestination({
        currentTripName: name,
        origin: originLocation,
        destination: destinationLocation,
      })
      dispatch(
        addTripToList({
          name: name,
          origin: originLocation,
          destination: destinationLocation,
        })
      )
      dispatch(hideDialog())
    }
  }
  return (
    <AddCommuteDialogWrapper {...props}>
      {displayView === 'IntroView' && (
        <IntroView goToChooseDestionationView={goToChooseDestinationView} />
      )}
      {displayView === 'ChooseDestinationView' && (
        <ChooseDestionationView
          goBack={() => {
            isIntroView ? goToIntroView() : dispatch(hideDialog())
          }}
          goToNameCommuteView={goToNameCommuteView}
          originLocation={originLocation}
          setOriginLocation={setOriginLocation}
          destinationLocation={destinationLocation}
          setDestinationLocation={setDestinationLocation}
        />
      )}
      {displayView === 'NameCommuteView' && (
        <NameCommuteView
          name={name}
          setName={setName}
          createTrip={createTrip}
          goToChooseDestionationView={goToChooseDestinationView}
        />
      )}
    </AddCommuteDialogWrapper>
  )
}

export default AddCommuteDialog
