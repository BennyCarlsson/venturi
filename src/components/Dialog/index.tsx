import { useAppDispatch } from 'hooks/redux'
import { useEffect } from 'react'
import { hideDialog, selectShowDialog } from 'redux/dialogSlice'
import { DialogWrapper, InnerWrapper } from './styles'
import { useAppSelector } from 'hooks/redux'
import AddCommuteDialog from './AddCommuteDialog'
import { AnimatePresence } from 'framer-motion'

type DialogProps = {}

const Dialog = (props: DialogProps) => {
  const dispatch = useAppDispatch()
  const displayDialog = useAppSelector(selectShowDialog)

  useEffect(() => {
    if (!displayDialog) return

    const handleOnKeyDown = (e: KeyboardEvent) => {
      if (e.key === 'Escape') {
        dispatch(hideDialog())
      }
    }

    document.addEventListener('keydown', handleOnKeyDown, false)

    return () => document.removeEventListener('keydown', handleOnKeyDown, false)
  }, [displayDialog, dispatch])

  if (!displayDialog) return null

  return (
    <AnimatePresence>
      {displayDialog && (
        <DialogWrapper {...props} data-testid='dialog-wrapper'>
          <InnerWrapper
            initial={{ opacity: 0 }}
            animate={{ opacity: 1 }}
            exit={{ opacity: 0 }}
          >
            <AddCommuteDialog />
          </InnerWrapper>
        </DialogWrapper>
      )}
    </AnimatePresence>
  )
}

export default Dialog
