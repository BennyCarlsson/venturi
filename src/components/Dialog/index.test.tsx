import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Dialog from '.';
import userEvent from '@testing-library/user-event';
import { hideDialog } from 'redux/dialogSlice';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { useSelector, useDispatch } from 'react-redux';

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: jest.fn()
}));

describe('<Dialog />', () => {
  const useAppDispatch = jest.fn();
  beforeEach(() => {
    (useSelector as jest.Mock).mockReturnValue({ dialog: { show: true } });
    (useDispatch as jest.Mock).mockReturnValue(useAppDispatch);
  });

  it('should match snapshot', () => {
    const tree = renderer.create(<Dialog />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render without crashing', () => {
    render(<Dialog />);
    expect(screen).toBeTruthy();
  });

  it('should call dispatch with hideDialog onClick', () => {
    render(<Dialog />);
    userEvent.click(screen.getByRole('button'));
    expect(useAppDispatch).toHaveBeenCalledWith(hideDialog());
    expect(useAppDispatch).toHaveBeenCalledTimes(1);
  });

  it('should call dispatch with hideDialog on Escape', () => {
    render(<Dialog />);
    const dialog = screen.getByTestId('dialog-wrapper');
    fireEvent.keyDown(dialog, { key: 'Escape', code: 'Escape' });
    expect(useAppDispatch).toHaveBeenCalledWith(hideDialog());
    expect(useAppDispatch).toHaveBeenCalledTimes(1);
  });
});
