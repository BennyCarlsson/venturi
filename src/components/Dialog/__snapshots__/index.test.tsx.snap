// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dialog /> should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfBwQ ibXtcp"
    data-testid="dialog-wrapper"
  >
    <div
      class="sc-gsTCUz jNtGXC"
      style="opacity: 0;"
    >
      <div
        class="sc-dIUggk hduPgm"
      >
        <span>
          <div
            class="sc-bqyKva fItXhb"
          >
            <div
              class="sc-kstrdz ievoUe"
            >
              <button
                class="sc-crrsfI heshZZ sc-idOhPF"
                data-testid="close-dialog-button"
              >
                <svg
                  class="sc-dQppl fUdGeY"
                >
                  chevron-down.svg
                </svg>
              </button>
              <h2
                class="sc-fubCfw hYsKMV sc-fodVxV hUHUQe"
                font-size="28"
              >
                Destinations
              </h2>
            </div>
            <p
              class="sc-dlfnbm cvhbVq"
              font-size="14"
            >
              Välj destinationer för din resa. Du kan alltid ändra eller lägga till flera vid ett senare tillfälle.
            </p>
          </div>
          <p
            class="sc-dlfnbm dJtete"
            font-size="18"
          >
            Från
          </p>
          <div
            class="sc-fFubgz ifWsfu"
          >
            <input
              class="sc-jSgupP begQqE"
              data-testid="from-input"
              value=""
            />
          </div>
          <p
            class="sc-dlfnbm dJtete sc-hBEYos iTMrrR"
            font-size="18"
          >
            Till
          </p>
          <div
            class="sc-fFubgz ifWsfu"
          >
            <input
              class="sc-jSgupP begQqE"
              data-testid="to-input"
              value=""
            />
          </div>
        </span>
        <button
          class="sc-gKsewC joGfoR sc-bkzZxe lhqPIK"
        >
          Namnge din resa
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
